File: ex/lib/gralkor/interpret.ex
defmodule Gralkor.Interpret do
  @moduledoc """
  Filter retrieved graph facts down to those relevant to the conversation,
  using the configured LLM.

  Two responsibilities, each its own tree:

    * `build_interpretation_context/3` — pure: assemble the LLM prompt from
      conversation messages and a formatted facts string, dropping oldest
      messages until the prompt fits the configured char budget. Renders
      role labels using `agent_name`.
    * `interpret_facts/4` — call the LLM with that prompt and a structured-
      output schema; return the list of relevant facts the LLM selected.

  See `ex-interpret` and `ex-interpret-context` in `gralkor/TEST_TREES.md`.
  """

  alias Gralkor.Message

  @default_budget 8_000

  @type interpret_fn :: (String.t() -> {:ok, [String.t()]} | {:error, term()})

  @doc """
  Run the LLM over the conversation context + facts text, returning the
  filtered list of relevant facts.

  Raises if the LLM call returns `{:error, _}` or a non-list response.
  Raises if `agent_name` is missing or blank.
  """
  @spec interpret_facts([Message.t()], String.t(), interpret_fn(), String.t(), keyword()) ::
          [String.t()]
  def interpret_facts(messages, facts_text, interpret_fn, agent_name, opts \\ [])
      when is_list(messages) and is_binary(facts_text) and is_function(interpret_fn, 1) do
    raise_if_blank!(agent_name)
    prompt = build_interpretation_context(messages, facts_text, agent_name, opts)

    case interpret_fn.(prompt) do
      {:ok, list} when is_list(list) ->
        list

      {:error, reason} ->
        raise "interpret failed: #{inspect(reason)}"

      other ->
        raise "interpret returned malformed response: #{inspect(other)}"
    end
  end

  @doc """
  Schema for the structured-output response the LLM returns.
  """
  @spec interpret_schema() :: keyword()
  def interpret_schema do
    [
      relevantFacts: [
        type: {:list, :string},
        required: true,
        doc:
          "Each entry is one fact line copied verbatim from the input " <>
            "(preserving every timestamp parenthetical such as '(created …)', " <>
            "'(valid from …)', '(invalid since …)', '(expired …)'; dropping the " <>
            "leading '- '), followed by ' — ' and a one-sentence relevance reason."
      ]
    ]
  end

  @doc """
  Assemble the LLM prompt from conversation messages and the formatted facts.

  Drops oldest messages until the assembled prompt fits the char budget
  (`opts[:budget]`, default #{@default_budget}). Raises on blank agent_name.
  """
  @spec build_interpretation_context([Message.t()], String.t(), String.t(), keyword()) ::
          String.t()
  def build_interpretation_context(messages, facts_text, agent_name, opts \\ [])
      when is_list(messages) and is_binary(facts_text) do
    raise_if_blank!(agent_name)
    budget = Keyword.get(opts, :budget, @default_budget)

    messages
    |> labelled_lines(agent_name)
    |> fit_to_budget(facts_text, budget)
    |> assemble(facts_text)
  end

  # ── internal ────────────────────────────────────────────────

  defp raise_if_blank!(name) when is_binary(name) do
    if String.trim(name) == "" do
      raise ArgumentError, "agent_name must be a non-blank string, got #{inspect(name)}"
    end

    :ok
  end

  defp raise_if_blank!(other) do
    raise ArgumentError, "agent_name must be a non-blank string, got #{inspect(other)}"
  end

  defp labelled_lines(messages, agent_name) do
    messages
    |> Enum.map(fn m -> {m.role, String.trim(m.content)} end)
    |> Enum.reject(fn {_, c} -> c == "" end)
    |> Enum.map(fn {role, content} -> render_line(role, content, agent_name) end)
  end

  defp render_line("user", content, _agent), do: "User: #{content}"
  defp render_line("assistant", content, agent), do: "#{agent}: #{content}"
  defp render_line("behaviour", content, agent), do: "#{agent}: (behaviour: #{content})"

  defp fit_to_budget([], _facts, _budget), do: []

  defp fit_to_budget(lines, facts, budget) do
    if String.length(assemble(lines, facts)) <= budget do
      lines
    else
      [_oldest | rest] = lines
      fit_to_budget(rest, facts, budget)
    end
  end

  defp assemble(lines, facts_text) do
    "Conversation context:\n" <>
      Enum.join(lines, "\n") <>
      "\n\nMemory facts to interpret:\n" <>
      facts_text
  end
end

File: ts/server/pipelines/interpret.py
from __future__ import annotations

from typing import TYPE_CHECKING

from pydantic import BaseModel, Field

from .messages import Message

if TYPE_CHECKING:
    from graphiti_core.llm_client import LLMClient


INTERPRET_SYSTEM_PROMPT = (
    "You are reviewing recalled memory facts for an agent mid-conversation. "
    "Each input fact is one line beginning with '- ' and may carry one or more "
    "timestamp parentheticals such as '(created …)', '(valid from …)', "
    "'(invalid since …)', '(expired …)'.\n\n"
    "Return only the facts that bear on the current task. For each one, produce "
    "a single string built from two parts joined by ' — ':\n"
    "  1. The original fact copied verbatim WITHOUT the leading '- '. Preserve "
    "every timestamp parenthetical exactly as given. Do not paraphrase, "
    "summarise, merge facts, drop timestamps, or reformat them.\n"
    "  2. One short sentence explaining why this fact is relevant to the "
    "current task.\n\n"
    "Example output entry: "
    "'Alice works at Acme (valid from 2024-01-01) (expired 2025-06-01) — "
    "confirms her former employer, which the user just asked about.'\n\n"
    "Skip facts with no bearing on the current task. If no facts are relevant, "
    "return an empty list. Do not return prose, prefixes, bullets, numbering, "
    "or any wrapping object — only the list of strings in the schema."
)

INTERPRET_TOKEN_BUDGET = 250_000
_CHARS_PER_TOKEN = 4
INTERPRET_CHAR_BUDGET = INTERPRET_TOKEN_BUDGET * _CHARS_PER_TOKEN

DEFAULT_OUTPUT_TOKEN_BUDGET = 2000


class InterpretParseFailed(Exception):
    """Raised when the LLM's interpret response cannot be parsed against the
    InterpretResult schema — typically because output_token_budget was too
    small and the response truncated mid-list. Distinct from generic
    RuntimeError so callers can catch parse failure specifically."""

    def __init__(self, message: str, raw_response: object = None) -> None:
        super().__init__(message)
        self.raw_response = raw_response


class InterpretResult(BaseModel):
    relevantFacts: list[str] = Field(
        description=(
            "List of relevant facts. Each entry is the original fact line "
            "copied verbatim (without the leading '- ', preserving every "
            "timestamp parenthetical such as '(valid from …)', '(invalid "
            "since …)', '(expired …)', '(created …)'), followed by ' — ' "
            "and one short sentence explaining why this fact is relevant. "
            "Empty list if nothing is relevant. No prose, no bullets, no "
            "numbering."
        )
    )


def _require_agent_name(agent_name: str) -> None:
    if agent_name is None or not str(agent_name).strip():
        raise ValueError("agent_name is required and must be non-blank")


def _require_positive_int(name: str, value: object) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
        raise ValueError(f"{name} must be a positive integer, got {value!r}")
    return value


def _render_label(role: str, agent_name: str) -> str:
    if role == "user":
        return "User"
    if role == "assistant":
        return agent_name
    if role == "behaviour":
        return agent_name
    return role.capitalize()


def _budget_instruction(output_token_budget: int) -> str:
    return (
        f"Respond within {output_token_budget} tokens. Keep each relevance "
        f"reason short so the full list fits."
    )


def build_interpretation_context(
    messages: list[Message],
    facts_text: str,
    agent_name: str,
    char_budget: int = INTERPRET_CHAR_BUDGET,
) -> str:
    _require_agent_name(agent_name)
    lines: list[str] = []
    for msg in messages:
        text = msg.content.strip()
        if not text:
            continue
        if msg.role == "behaviour":
            lines.append(f"{agent_name}: (behaviour: {text})")
        else:
            lines.append(f"{_render_label(msg.role, agent_name)}: {text}")

    budget = char_budget
    trimmed: list[str] = []
    for line in reversed(lines):
        if budget <= 0:
            break
        trimmed.insert(0, line)
        budget -= len(line)

    return (
        "Conversation context:\n"
        + "\n".join(trimmed)
        + "\n\nMemory facts to interpret:\n"
        + facts_text
    )


async def interpret_facts(
    messages: list[Message],
    facts_text: str,
    llm_client: "LLMClient",
    agent_name: str,
    output_token_budget: int = DEFAULT_OUTPUT_TOKEN_BUDGET,
) -> list[str]:
    _require_agent_name(agent_name)
    _require_positive_int("output_token_budget", output_token_budget)
    if llm_client is None:
        raise RuntimeError(
            "interpret_facts: llm_client is required (configure an LLM provider API key)"
        )

    from graphiti_core.prompts.models import Message as LLMMessage

    context = build_interpretation_context(messages, facts_text, agent_name)
    context_with_budget = context + "\n\n" + _budget_instruction(output_token_budget)
    prompt = [
        LLMMessage(role="system", content=INTERPRET_SYSTEM_PROMPT),
        LLMMessage(role="user", content=context_with_budget),
    ]

    response = await llm_client.generate_response(
        prompt,
        response_model=InterpretResult,
        max_tokens=output_token_budget,
    )
    if not isinstance(response, dict):
        raise InterpretParseFailed(
            "interpret_facts: response was not a dict (schema mismatch or truncation)",
            raw_response=response,
        )
    raw = response.get("relevantFacts")
    if not isinstance(raw, list):
        raise InterpretParseFailed(
            "interpret_facts: relevantFacts missing or not a list (schema mismatch or truncation)",
            raw_response=response,
        )
    return [str(item).strip() for item in raw if str(item).strip()]
